src/lib.rs

```diff
@@ -505,6 +526,49 @@ where
     }
 }
 
+impl<T, Q, N> Index<&'_ Q> for Counter<T, N>
+where
+    T: Hash + Eq + Borrow<Q>,
+    Q: Hash + Eq,
+    N: Zero,
+{
+    type Output = N;
+
+    /// Index in immutable contexts
+    ///
+    /// ```
+    /// # use counter::Counter;
+    /// let counter = Counter::<_>::init("aabbcc".chars());
+    /// assert_eq!(counter[&'a'], 2);
+    /// assert_eq!(counter[&'b'], 2);
+    /// assert_eq!(counter[&'c'], 2);
+    /// assert_eq!(counter[&'d'], 0);
+    /// ```
+    fn index(&self, key: &'_ Q) -> &N {
+        self.map.get(key).unwrap_or(&self.zero)
```

Comment:
Why use `self.zero` here? It looks like this would also work, for a smaller struct size:

```suggestion
        self.map.get(key).unwrap_or(&N::zero())
```

Reply:
No. The `zero` must be a struct field or `static` variable. And in the current Rust, we cannot declare generic `static` variables. Moreover `Zero::zero` is not a `const fn`.

```
error[E0515]: cannot return value referencing temporary value
   --> src/lib.rs:548:9
    |
548 |         self.map.get(key).unwrap_or(&N::zero())
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^---------^
    |         |                            |
    |         |                            temporary value created here
    |         returns a value referencing data owned by the current function

error: aborting due to previous error

For more information about this error, try `rustc --explain E0515`.
```

As an alternative, we can use `RefCell` to modify the inner `HashMap` for `&self`.